### PR TITLE
[DeepSeek-V4 CSA 2/4] Add csa_compressor op and compressor module

### PR DESCRIPTION
**Pull request by ddjekicTT**
## Summary
Second of the CSA stack. Adds the `csa_compressor` device op and the Python compressor module that
drives it.
Issue: https://github.com/tenstorrent/tt-metal/issues/54456

`ttnn.experimental.deepseek_prefill.csa_compressor(kv, gate, position_bias, initial_kv_state,
initial_score_state, seq_len_actual, ...)` performs the windowed pooling + Blaze-parity state update and
internally calls `compressor_state_exchange` (PR 1) to hand SP state across ranks.

### How compression is done
CSA pooling at compress rate 4 with a one-window overlap. Every group of 4 tokens ("window") produces one pooled row, but the softmax pool runs over 8 candidates: 4 from the previous window and 4 from the current one.

That's why kv and gate are [1, 1, S_local, 2*head_dim] — the left half of the width is the Ca projection (contributed to the next window) and the right half is Cb (contributed to its own window). head_dim is derived as width / 2.

### Top level: `ttnn.experimental.deepseek_prefill.csa_compressor()`

image

### Compression (per output tile)
image

### Compressor stacked perf on bh mesh 2x2 (640 local seq len)
image

## Contents
- **C++ op** `csa_compressor/`: host API, device operation, program factory, and compute/reader/writer/
  state-update kernels; bound via `experimental_nanobind.cpp`.
- **`tt/mla/compressor.py`**: `TtCompressorUtils`, `TtCompressorBase`, `TtHCACompressor`, `TtCSACompressor`,
  plus the host-side state-layout helpers (`CSA_STATE_ROWS`, `csa_state_rows`, `csa_state_predecessor_ca`,
  `csa_slab_align`, `rope_table_tokens`).
- **`tt/tt_ccl.py`**: `resolve_per_axis_topology`, the shared per-axis (SP/TP) topology normalizer used by
  the compressors and V4 attention blocks.
- **Shared test infra**: new `fabric_profiles` helpers (plain-1D / disabled) and `mesh_configs.V4_MESH_CONFIGS`.
- **Tests**: `test_csa_compressor.py` (op unit) and `test_ttnn_csa_compressor.py` (PCC).

## Stack
- 2/4 — base PR 1 (compressor_state_exchange)

## Test
`pytest models/demos/deepseek_v3_d_p/tests/op_unit_tests/test_csa_compressor.py`
`pytest models/demos/deepseek_v3_d_p/tests/pcc/test_ttnn_csa_compressor.py`

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ddjekic/csa-2-compressor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ddjekic/csa-2-compressor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ddjekic/csa-2-compressor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ddjekic/csa-2-compressor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ddjekic/csa-2-compressor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ddjekic/csa-2-compressor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ddjekic/csa-2-compressor)

### CODE REVIEW

**Review by dpopovTT — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]

File: models/demos/deepseek_v3_d_p/tests/pcc/test_ttnn_csa_compressor.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+import ttnn
+from models.demos.deepseek_v3_d_p.reference.deepseek_v4.modeling_deepseek_v4 import (
+    DeepseekV4CSACompressor,
+    apply_rotary_pos_emb,
+)
+from models.demos.deepseek_v3_d_p.reference.deepseek_v4_flash_config import DeepSeekV4FlashConfig
+from models.demos.deepseek_v3_d_p.tests.op_unit_tests.test_csa_compressor import _torch_csa_compressor
+from models.demos.deepseek_v3_d_p.tests.pcc.mesh_configs import V4_MESH_CONFIGS
+from models.demos.deepseek_v3_d_p.tests.pcc.test_ttnn_csa import _SEED, _config
```
Comment (jmitrovicTT) [inline models/demos/deepseek_v3_d_p/tests/pcc/test_ttnn_csa_compressor.py]: This looks like a blocker: `test_ttnn_csa.py` doesn't exist on this branch...

this test can't be run on this branch:
```ERROR: file or directory not found: tt-metal/models/demos/deepseek_v3_d_p/tests/pcc/test_ttnn_csa_compressor.py```

File: models/demos/deepseek_v3_d_p/tests/pcc/test_ttnn_csa_compressor.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+from models.demos.deepseek_v3_d_p.tt.mla.compressor import CSA_STATE_ROWS, TtCSACompressor
+from tests.ttnn.utils_for_testing import assert_with_pcc
+
+# PER-CHIP padded prompt length, not global, matching tests/pcc/test_ttnn_csa.py.
+_LOCAL_SHAPES = [640]
+# How far short of a whole slab the real prompt stops. prepare_input pads it straight back up, so the
+# per-chip shape is exactly _LOCAL_SHAPES while the last rank still ends mid-window.
+_RAGGED_TAIL = 2
```
Comment (jmitrovicTT) [inline models/demos/deepseek_v3_d_p/tests/pcc/test_ttnn_csa_compressor.py]: Every run here is ragged, the aligned case (tail 0), which is the only one prefill actually produces, isn't covered at all. It makes sense for `_RAGGED_TAIL`  to maki  it `[0, 2]` with `aligned`/`ragged` ids for example

File: models/demos/deepseek_v3_d_p/tt/mla/compressor.py [hunk 3]
```diff
@@ -174,76 +249,88 @@ def prepare_input(hidden: torch.Tensor, sp_factor: int, compress_rate: int):
             hidden = torch.nn.functional.pad(hidden, (0, 0, 0, pad))
         return hidden, seq_len_actual
 
-
-class TtHCACompressor(TtCompressorBase):
```
Comment (jmitrovicTT) [inline models/demos/deepseek_v3_d_p/tt/mla/compressor.py]: Pls run Blaze CI for hca_chunked, hca_perf

File: models/demos/deepseek_v3_d_p/tests/pcc/test_ttnn_csa_compressor.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+    seq_len = local_seq_len * mesh_device.shape[0] - _RAGGED_TAIL
+    hidden = torch.randn(1, seq_len, config.hidden_size)
+    compress_rate = config.compress_rates["compressed_sparse_attention"]
+    hidden_padded, seq_len_actual = TtCSACompressor.prepare_input(hidden, mesh_device.shape[0], compress_rate)
+    head_dim = config.head_dim
+    initial_kv = torch.zeros(1, 1, CSA_STATE_ROWS, head_dim, dtype=torch.bfloat16)
+    initial_score = torch.full_like(initial_kv, float("-inf"))
+    with torch.no_grad():
+        expected, expected_kv_state, expected_score_state = _golden(
```
Comment (jmitrovicTT) [inline models/demos/deepseek_v3_d_p/tests/pcc/test_ttnn_csa_compressor.py]: `_golden` uses `_torch_csa_compressor` for the compression, so nothing in this PR checks the outputs against the actual reference (DeepseekV4HCACompressor)...

The state genuinely has no reference counterpart, so keeping that on `_torch_csa_compressor` makes sense. But `compressed_kv` does... Could you add that as a PCC check

**Review by jmitrovicTT — CHANGES_REQUESTED** [review]
